[HDR] Support HDR snapshotting for WKThumbnailView.

mattwoodrow opened the pull request:
#### 1f8b191c3bd65655a5a8d67e025c69b701f7dbb1
<pre>
[HDR] Support HDR snapshotting for WKThumbnailView.
<a href="https://bugs.webkit.org/show_bug.cgi?id=283316">https://bugs.webkit.org/show_bug.cgi?id=283316</a>
&lt;<a href="https://rdar.apple.com/140144773">rdar://140144773</a>&gt;

Reviewed by Simon Fraser.

Converts WebPageProxy::takeSnapshot to return a CGImage, and moves the
complexity of unpacking an ImageBufferBackendHandle from WKThumbailView into
WebPageProxy. Moves all consumers of takeSnapshotLegacy to the new version,
since they all just wanted a CGImage already.

Add SnapshotOption::AllowHDR which makes takeSnapshot potentially return a
RGBA16F surface (with extended range color space) if the web page includes HDR
content. Returns the painted headroom of the content back to WebPageProxy, and
stores this on the output CGImage from takeSnapshot.

Opt-in to HDR snapshots for WKThumbnailView, and configure the CALayer with the
headroom and tonemap requests if we get an HDR result.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::drawsHDRContent const):
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/graphics/DestinationColorSpace.h:
* Source/WebCore/platform/graphics/ImageTypes.h:
(WebCore::Headroom::operator&lt;):
(IPC::AsyncReplyError&lt;WebCore::Headroom&gt;::create):
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::IOSurface::sinkIntoImage):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/ImageOptions.h:
(WebKit::snapshotOptionsToImageOptions):
* Source/WebKit/Shared/ImageOptions.serialization.in:
* Source/WebKit/Shared/WebImage.cpp:
(WebKit::WebImage::create):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView takeSnapshotWithConfiguration:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKThumbnailView.mm:
(-[_WKThumbnailView requestSnapshot]):
(-[_WKThumbnailView _didTakeSnapshot:]):
(-[_WKThumbnailView _didTakeSnapshot:size:]): Deleted.
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _snapshotRectAfterScreenUpdates:rectInViewCoordinates:intoImageOfWidth:completionHandler:]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::takeSnapshotLegacy):
(WebKit::WebPageProxy::takeSnapshot):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WKTextAnimationManagerMac.mm:
(-[WKTextAnimationManager textPreviewForRect:completion:]):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::takeSnapshot):
(WebKit::snapshotColorSpace):
(WebKit::WebPage::snapshotAtSize):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/297373@main">https://commits.webkit.org/297373@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bcd49b30c20d2ba37d0abdf6c5a89b201dc7d905

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111555 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31222 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21686 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117590 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/61826 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113517 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31902 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39803 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84766 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/61826 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114502 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25461 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100412 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65207 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/110988 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24814 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18550 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61418 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/94857 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18617 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/120841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38608 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28686 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/120841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38985 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96674 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/120841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23828 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38626 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16416 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/34644 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38497 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43970 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/38160 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41496 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39862 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/Page.cpp:
```diff
@@ -5840,6 +5840,16 @@ bool Page::requiresUserGestureForVideoPlayback() const
+    bool drawsHDRContent = false;
+    forEachDocument([&] (Document& document) {
+        if (document.drawsHDRContent())
+            drawsHDRContent = true;
```
If we could break here it would be a little faster. Maybe `forEachRenderableDocument` ?

smfr commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -13714,16 +13714,50 @@ void WebPageProxy::setScrollPerformanceDataCollectionEnabled(bool enabled)
+        RetainPtr<CGImageRef> image;
+        WTF::switchOn(*imageHandle,
+            [&image] (WebCore::ShareableBitmap::Handle& handle) {
+                if (auto bitmap = WebCore::ShareableBitmap::create(WTFMove(handle), WebCore::SharedMemory::Protection::ReadOnly))
```
`if (RefPtr bitmap = WebCore...`

smfr commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -1948,7 +1948,9 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     void takeSnapshotLegacy(WebCore::IntRect, WebCore::IntSize bitmapSize, SnapshotOptions, CompletionHandler<void(std::optional<WebCore::ShareableBitmapHandle>&&)>&&);
-    void takeSnapshot(WebCore::IntRect, WebCore::IntSize bitmapSize, SnapshotOptions, CompletionHandler<void(std::optional<ImageBufferBackendHandle>&&)>&&);
+#if PLATFORM(COCOA)
+    void takeSnapshot(WebCore::IntRect, WebCore::IntSize bitmapSize, SnapshotOptions, CompletionHandler<void(CGImageRef)>&&);
```
Not you, but we should pass `WebCore::IntRect` by reference in these functions.

smfr commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -3044,55 +3044,63 @@ void WebPage::setFooterBannerHeight(int height)
     RefPtr coreFrame = m_mainFrame->coreLocalFrame();
     if (!coreFrame) {
-        completionHandler(WTFMove(handle));
+        completionHandler(WTFMove(handle), 1);
```
Maybe we should have a `const float SDRHeadroom = 1.0f` somewhere.

mattwoodrow replied:
Switched from `float` to `WebCore::Headroom` so that we can use the `Headroom::None` constant.

smfr commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -3178,28 +3186,47 @@ void WebPage::paintSnapshotAtSize(const IntRect& rect, const IntSize& bitmapSize
 {
 #if ENABLE(PDF_PLUGIN)
-    imageOptions.add(m_pluginViews.computeSize() ? ImageOption::Local : ImageOption::Shareable);
+    ImageOptions imageOptions = m_pluginViews.computeSize() ? ImageOption::Local : ImageOption::Shareable;
```
auto

mattwoodrow replied:
`auto` constructs an `ImageOption` (enum value), but we need `ImageOptions` (`OptionSet<ImageOption>`).

smfr commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -3178,28 +3186,47 @@ void WebPage::paintSnapshotAtSize(const IntRect& rect, const IntSize& bitmapSize
+    ImageOptions imageOptions = m_pluginViews.computeSize() ? ImageOption::Local : ImageOption::Shareable;
 #else
-    imageOptions.add(ImageOption::Shareable);
+    ImageOptions imageOptions = ImageOption::Shareable;
```
auto

smfr commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -2248,7 +2248,7 @@ class WebPage final : public API::ObjectImpl<API::Object::Type::BundlePage>, pub
     void getAccessibilityTreeData(CompletionHandler<void(const std::optional<IPC::SharedBufferReference>&)>&&);
     void updateRenderingWithForcedRepaint(CompletionHandler<void()>&&);
-    void takeSnapshot(WebCore::IntRect snapshotRect, WebCore::IntSize bitmapSize, SnapshotOptions, bool, CompletionHandler<void(std::optional<ImageBufferBackendHandle>&&)>&&);
+    void takeSnapshot(WebCore::IntRect snapshotRect, WebCore::IntSize bitmapSize, SnapshotOptions, CompletionHandler<void(std::optional<ImageBufferBackendHandle>&&, float)>&&);
```
Please name the float parameter (it's very non-obvious from here).

mattwoodrow committed:
[HDR] Support HDR snapshotting for WKThumbnailView.
https://bugs.webkit.org/show_bug.cgi?id=283316
<rdar://140144773>

Reviewed by Simon Fraser.

Converts WebPageProxy::takeSnapshot to return a CGImage, and moves the
complexity of unpacking an ImageBufferBackendHandle from WKThumbailView into
WebPageProxy. Moves all consumers of takeSnapshotLegacy to the new version,
since they all just wanted a CGImage already.

Add SnapshotOption::AllowHDR which makes takeSnapshot potentially return a
RGBA16F surface (with extended range color space) if the web page includes HDR
content. Returns the painted headroom of the content back to WebPageProxy, and
stores this on the output CGImage from takeSnapshot.

Opt-in to HDR snapshots for WKThumbnailView, and configure the CALayer with the
headroom and tonemap requests if we get an HDR result.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::drawsHDRContent const):
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/graphics/DestinationColorSpace.h:
* Source/WebCore/platform/graphics/ImageTypes.h:
(WebCore::Headroom::operator<):
(IPC::AsyncReplyError<WebCore::Headroom>::create):
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::IOSurface::sinkIntoImage):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/ImageOptions.h:
(WebKit::snapshotOptionsToImageOptions):
* Source/WebKit/Shared/ImageOptions.serialization.in:
* Source/WebKit/Shared/WebImage.cpp:
(WebKit::WebImage::create):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView takeSnapshotWithConfiguration:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKThumbnailView.mm:
(-[_WKThumbnailView requestSnapshot]):
(-[_WKThumbnailView _didTakeSnapshot:]):
(-[_WKThumbnailView _didTakeSnapshot:size:]): Deleted.
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _snapshotRectAfterScreenUpdates:rectInViewCoordinates:intoImageOfWidth:completionHandler:]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::takeSnapshotLegacy):
(WebKit::WebPageProxy::takeSnapshot):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WKTextAnimationManagerMac.mm:
(-[WKTextAnimationManager textPreviewForRect:completion:]):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::takeSnapshot):
(WebKit::snapshotColorSpace):
(WebKit::WebPage::snapshotAtSize):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/297373@main
